src/tcpconns.c

```diff
@@ -318,6 +341,137 @@ static uint32_t count_total[TCP_STATE_MAX + 1];
+  } else if (rc < 0) {
+    ERROR("[I; crash] Cache insert failed for %s", key);
+    sfree(key);
+    counter_cache_free_entry(*entry);
+    *entry = NULL;
+    goto exit;
+  }
+exit:
+  pthread_mutex_unlock(&counter_cache_lock);
+  return ret;
+}
+
+/* Cleans up cache by removing entries last updated more than
+ * counter_cache_timeout before t.  Returns number of elements removed. */
+static size_t counter_cache_cleanup(cdtime_t t) {
+  size_t i;
+  char **remove_keys = NULL;
+  size_t remove_keys_num = 0;
+  size_t remove_keys_size = 0;
+  c_avl_iterator_t *iter;
+  char *key;
+  counter_cache_entry_t entry;
+  char t_str[32];
+
+  cdtime_to_iso8601(t_str, sizeof(t_str), t);
+  INFO("counter_cache_cleanup: Clean up @%s", t_str);
+
+  pthread_mutex_lock(&counter_cache_lock);
+  iter = c_avl_get_iterator(counter_cache_tree);
+  while (c_avl_iterator_next(iter, (void**)&key, (void**)&entry) == 0) {
+    if (entry->last_update_time < t - counter_cache_timeout) {
+      cdtime_to_iso8601(t_str, sizeof(t_str), entry->last_update_time);
+      DEBUG("counter_cache_cleanup: Remove entry for %s (last update @%s)",
+            t_str);
+
+      if (remove_keys_size < remove_keys_num + 1) {
+        remove_keys_size *= 2;
+        if (remove_keys_size < remove_keys_num + 1)
+          remove_keys_size = remove_keys_num + 1;
+        remove_keys = realloc(
+            remove_keys, remove_keys_size * sizeof(*remove_keys));
+      }
+
+      remove_keys[remove_keys_num++] = key;
+    }
+  }
+
+  for (i = 0; i < remove_keys_num; i++) {
+    entry = NULL;
+    if (!c_avl_remove(counter_cache_tree, remove_keys[i],
+                      NULL, (void**)&entry)) {
+      ERROR("[I] Key-to-remove %s no longer in tree", remove_keys[i]);
+    } else if (entry) {
+      sfree(remove_keys[i]);
+      sfree(entry);
+    }
+  }
+
+  pthread_mutex_unlock(&counter_cache_lock);
+
+  INFO("Cleaned up %zu keys", remove_keys_num);
+  return remove_keys_num;
+}
+
+static void *call_counter_cache_cleanup(void *unused) {
+  while (!shutdown_module) {
+    counter_cache_cleanup(cdtime());
+    sleep(CDTIME_T_TO_DOUBLE(counter_cache_timeout) / 4 + 1);
+  }
+  return NULL;
+}
+
+static uint64_t get_max_for_field_type(enum tcpi_field_type type) {
+  switch(type) {
+    case UINT8: return 1ULL << 8;
+    case UINT32: return 1ULL << 32;
+    case UINT64: return 0;  /* Arithmetic works when adding and subtracting. */
+    default: return 0;      /* Impossible, but this value might be good. */
+  }
+}
+
+/* Returns a wraparound-adjusted delta for going from old to new in field of
+ * given type. Values > counter_max_delta are ignored and return 0. */
+static uint64_t get_wraparound_delta(
+    enum tcpi_field_type type, uint64_t new, uint64_t old) {
+  uint64_t delta;
+  if (old <= new) {
+    /* Likely case: no wraparound */
+    delta = new - old;
+  }
+  else {
+    delta = get_max_for_field_type(type) - (old - new);
+  }
+  return delta <= counter_max_delta ? delta : 0;
```

Comment:
Agreed (on both counts). It's actually not that bad, there's a *lot* of flexibility in picking the value: it's mostly for when wraparound detection can't be trusted.